File: src/main/scala/com/yammer/dropwizard/jetty/NonblockingServletHolder.scala
package com.yammer.dropwizard.jetty

import org.eclipse.jetty.servlet.ServletHolder
import org.eclipse.jetty.server.Request
import javax.servlet.{ServletResponse, ServletRequest, Servlet}

/**
  * A {@link ServletHolder} subclass which removes much of the synchronization
  * around servlet initialization, preferring instead to simply use an
  * already-created servlet instance.
  */
class NonblockingServletHolder(servlet: Servlet) extends ServletHolder(servlet) {
  override def getServlet = servlet

  override def handle(baseRequest: Request, request: ServletRequest, response: ServletResponse) {
    val suspendable = baseRequest.isAsyncSupported
    if (!isAsyncSupported) {
      baseRequest.setAsyncSupported(false)
    }
    try {
      servlet.service(request, response)
    } finally {
      baseRequest.setAsyncSupported(suspendable)
    }
  }
}

File: dropwizard/src/main/java/com/yammer/dropwizard/jetty/NonblockingServletHolder.java
package com.yammer.dropwizard.jetty;

import org.eclipse.jetty.server.Request;
import org.eclipse.jetty.servlet.ServletHolder;

import javax.servlet.*;
import java.io.IOException;

/**
 * A {@link ServletHolder} subclass which removes the synchronization around servlet initialization
 * by requiring a pre-initialized servlet holder.
 */
public class NonblockingServletHolder extends ServletHolder {
    private final Servlet servlet;

    public NonblockingServletHolder(Servlet servlet) {
        super(servlet);
        this.servlet = servlet;
    }

    @Override
    public Servlet getServlet() throws ServletException {
        return servlet;
    }

    @Override
    public void handle(Request baseRequest,
                       ServletRequest request,
                       ServletResponse response) throws ServletException, IOException {
        final boolean asyncSupported = baseRequest.isAsyncSupported();
        if (!isAsyncSupported()) {
            baseRequest.setAsyncSupported(false);
        }
        try {
            servlet.service(request, response);
        } finally {
            baseRequest.setAsyncSupported(asyncSupported);
        }
    }
}
